pkg/controller/v1alpha2/applicationconfiguration/component.go

```diff
@@ -182,6 +206,63 @@ func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtim
 	return true
 }
 
+func (c *ComponentHandler) cleanupControllerRevision(mt metav1.Object, curComp *v1alpha2.Component) error {
+	labels := &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			ControllerRevisionComponentLabel: curComp.Name,
+		},
+	}
+	selector, err := metav1.LabelSelectorAsSelector(labels)
+	if err != nil {
+		return err
+	}
+	revisions, err := c.appsClient.ControllerRevisions(mt.GetNamespace()).List(context.Background(), metav1.ListOptions{
+		LabelSelector: selector.String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Get appConfigs and workloads filter controllerRevision used
+	appConfigs := &v1alpha2.ApplicationConfigurationList{}
+	if err := c.client.List(context.Background(), appConfigs); err != nil {
+		return err
+	}
+
+	liveHashes := make(map[string]bool)
+	for _, appConfig := range appConfigs.Items {
+		for _, component := range appConfig.Spec.Components {
+			if component.RevisionName != "" {
+				liveHashes[component.RevisionName] = true
+			}
+		}
+	}
+
+	toKeep := int(*curComp.Spec.RevisionLimit)
```

Comment:
I think `toKeep = revisionLimit + len(liveHashes)`

Reply:
Livehashes may before the limit or after the limit, if `toKeep = revisionLimit + len(liveHashes)`, the final count may bigger than `toKeep`, by the way if the hash is used, it will be skipped.

Reply:
RevisionHistory is used be roll back, if `len(liveHashes)` > revisionLimit, we will clean all rest no used revisions, which means we have no revision to roll back.

Reply:
![image](https://user-images.githubusercontent.com/16429864/87366756-08bef300-c5ac-11ea-8872-3bcc0f7f7469.png)
LiveHashes will skipped, don't worry cleaning used revisions.

Reply:
👌